storagenode/pieces/cache.go

```diff
@@ -0,0 +1,295 @@
+// estimate how many bytes missed then add those to the space used result of iteration.
+func (blobs *BlobsUsageCache) Recalculate(ctx context.Context, newTotal, totalAtIterationStart int64, newTotalBySatellite, totalBySatelliteAtIterationStart map[storj.NodeID]int64) error {
+	blobs.mu.Lock()
+	totalsAtIterationEnd := blobs.copyCacheTotals()
+	blobs.mu.Unlock()
+
+	estimatedTotals := estimate(newTotal,
+		totalAtIterationStart,
+		totalsAtIterationEnd.totalSpaceUsed,
+	)
+
+	var estimatedTotalsBySatellite = map[storj.NodeID]int64{}
+	for ID, newTotal := range newTotalBySatellite {
+		estimatedNewTotal := estimate(newTotal,
+			totalBySatelliteAtIterationStart[ID],
+			totalsAtIterationEnd.totalSpaceUsedBySatellite[ID],
+		)
+		// if the estimatedNewTotal is zero then there is no data stored
+		// for this satelliteID so don't add it to the cache
+		if estimatedNewTotal == 0 {
+			continue
+		}
+		estimatedTotalsBySatellite[ID] = estimatedNewTotal
+	}
+
+	// find any saIDs that are in totalsAtIterationEnd but not in newTotalSpaceUsedBySatellite
+	missedWhenIterationEnded := getMissed(totalsAtIterationEnd.totalSpaceUsedBySatellite,
+		newTotalBySatellite,
+	)
+	if len(missedWhenIterationEnded) > 0 {
+		for ID := range missedWhenIterationEnded {
+			estimatedNewTotal := estimate(0,
+				totalBySatelliteAtIterationStart[ID],
+				totalsAtIterationEnd.totalSpaceUsedBySatellite[ID],
+			)
+			if estimatedNewTotal == 0 {
+				continue
+			}
+			estimatedTotalsBySatellite[ID] = estimatedNewTotal
+		}
+	}
+
+	blobs.mu.Lock()
+	blobs.totalSpaceUsed = estimatedTotals
+	blobs.totalSpaceUsedBySatellite = estimatedTotalsBySatellite
+	blobs.mu.Unlock()
+	return nil
+}
+
+func estimate(newSpaceUsedTotal, totalAtIterationStart, totalAtIterationEnd int64) int64 {
+	if newSpaceUsedTotal == totalAtIterationEnd {
+		return newSpaceUsedTotal
+	}
+
+	// If we missed writes/deletes while iterating, we will assume that half of those missed occurred before
+	// the iteration and half occurred after. So here we add half of the delta to the result space used totals
+	// from the iteration to account for those missed.
+	spaceUsedDeltaDuringIteration := totalAtIterationEnd - totalAtIterationStart
+	estimatedTotal := newSpaceUsedTotal + (spaceUsedDeltaDuringIteration / 2)
+	if estimatedTotal < 0 {
+		return 0
+	}
+	return estimatedTotal
+}
+
+func getMissed(endTotals, newTotals map[storj.NodeID]int64) map[storj.NodeID]int64 {
+	var missed = map[storj.NodeID]int64{}
+	for id, total := range endTotals {
+		if _, ok := newTotals[id]; !ok {
+			missed[id] = total
+		}
+	}
+	return missed
+}
+
+// Close satisfies the pieces interface
+func (blobs *BlobsUsageCache) Close() error {
+	return nil
+}
+
+// TestCreateV0 creates a new V0 blob that can be written. This is only appropriate in test situations.
+func (blobs *BlobsUsageCache) TestCreateV0(ctx context.Context, ref storage.BlobRef) (_ storage.BlobWriter, err error) {
```

Comment:
is there any way to move this to a test specific file?

Reply:
i could move, but then i would need to import this package. i would prefer to leave here.